batcher/sweep_batch.go

```diff
@@ -0,0 +1,1330 @@
+	// If the feeRate is unset then we never published before, so we
+	// retrieve the fee estimate from our wallet.
+	if b.RbfCache.FeeRate == 0 {
+		b.log.Infof("initializing rbf fee rate for conf target=%v", b.Cfg.BatchConfTarget)
+		rate, err := b.wallet.EstimateFeeRate(
+			ctx, b.Cfg.BatchConfTarget,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Set the initial value for our fee rate.
+		b.RbfCache.FeeRate = rate
+	} else {
+		// Bump the fee rate by the configured step.
+		b.RbfCache.FeeRate += defaultFeeRateStep
+		b.RbfCache.LastHeight = b.currentHeight
+		return b.persist(ctx)
+	}
+
+	return nil
+}
+
+// monitorBlockHeight monitors the current block height.
+func (b *Batch) monitorBlockHeight(ctx context.Context) {
+	blockChan, errChan, err := b.chainNotifier.RegisterBlockEpochNtfn(
+		ctx,
+	)
+	if err != nil {
+		b.errChan <- err
+		return
+	}
+
+	for {
+		select {
+		case height := <-blockChan:
+			select {
+			case b.blockEpochChan <- height:
+			case <-ctx.Done():
+				return
+			}
+
+		case err := <-errChan:
+			select {
+			case b.errChan <- err:
+			case <-ctx.Done():
+				return
+			}
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// monitorSpend monitors the primary sweep's outpoint for spends. The reason we
+// monitor the primary sweep's outpoint is because the primary sweep was the
+// first sweep that entered this batch, therefore it is present in all the
+// versions of the batch transaction. This means that even if an older version
+// of the batch transaction gets confirmed, due to the uncertainty of RBF
+// replacements and network propagation, we can always detect the transaction.
+func (b *Batch) monitorSpend(ctx context.Context) {
+	// If no primary sweep exists the batch is empty and there is no spend
+	// to monitor.
+	if b.PrimarySweepID == lntypes.ZeroHash {
+		b.log.Warnf("can't monitor batch spend, no primary sweep")
+		return
+	}
+
+	// Generate this routine's nonce. This is used to signal to the previous
+	// routine that a new routine is now monitoring the spend.
+	randValue, err := crand.Int(crand.Reader, big.NewInt(math.MaxInt32))
+	if err != nil {
+		b.errChan <- err
+		return
+	}
+
+	nonce := int32(randValue.Int64())
+	// Send the nonce to the refresh channel.
+	b.spendRefreshChan <- nonce
+
+	// Get the primary sweep.
+	b.sweepsLock.Lock()
+	sweep := b.sweeps[b.PrimarySweepID]
+	b.sweepsLock.Unlock()
+
+	spendCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	spendChan, spendErr, err := b.chainNotifier.RegisterSpendNtfn(
+		spendCtx, &sweep.Outpoint, sweep.Htlc.PkScript,
+		sweep.InitiationHeight,
+	)
+	if err != nil {
+		select {
+		case b.errChan <- err:
+		case <-ctx.Done():
+		}
+
+		return
+	}
+
+	b.log.Infof("monitoring spend for outpoint %s", sweep.Outpoint.String())
+	for {
+		select {
+		case spend := <-spendChan:
+			select {
+			case b.spendChan <- spend:
+
+			case <-ctx.Done():
+			}
+
+			return
+
+		case err := <-spendErr:
+			select {
+			case b.errChan <- err:
```

Comment:
I'd recommend extracting a method, instead of having the select everywhere where we want to send an error to the main eventloop.